subiquitycore/tui.py

```diff
@@ -0,0 +1,279 @@
+            if ss.scripts:
+                self.aio_loop.call_soon(_run_script)
+
+        def c(pat):
+            but = view_helpers.find_button_matching(self.ui, '.*' + pat + '.*')
+            if not but:
+                ss.wait_count += 1
+                if ss.wait_count > 10:
+                    raise Exception("no button found matching %r after"
+                                    "waiting for 10 secs" % pat)
+                wait(1, func=lambda: c(pat))
+                return
+            ss.wait_count = 0
+            view_helpers.click(but)
+
+        def wait(delay, func=None):
+            ss.waiting = True
+
+            def next():
+                ss.waiting = False
+                if func is not None:
+                    func()
+                if not ss.waiting:
+                    ss.scripts = ss.scripts[1:]
+                    if ss.scripts:
+                        _run_script()
+            self.aio_loop.call_later(delay, next)
+
+        ss.ns['c'] = c
+        ss.ns['wait'] = wait
+        ss.ns['ui'] = self.ui
+
+        self.aio_loop.call_later(0.06, _run_script)
+
+    def toggle_rich(self):
+        if self.rich_mode:
+            urwid.util.set_encoding('ascii')
+            new_palette = PALETTE_MONO
+            self.rich_mode = False
+        else:
+            urwid.util.set_encoding('utf-8')
+            new_palette = PALETTE_COLOR
+            self.rich_mode = True
+        urwid.CanvasCache.clear()
+        self.urwid_loop.screen.register_palette(new_palette)
+        self.urwid_loop.screen.clear()
+
+    def unhandled_input(self, key):
+        if self.opts.dry_run and key == 'ctrl x':
+            self.exit()
+        elif key == 'f3':
+            self.urwid_loop.screen.clear()
+        elif self.opts.run_on_serial and key in ['ctrl t', 'f4']:
+            self.toggle_rich()
+
+    def extra_urwid_loop_args(self):
+        return {}
+
+    def make_screen(self, inputf=None, outputf=None):
+        return make_screen(self.opts.ascii, inputf, outputf)
+
+    def start_urwid(self, input=None, output=None):
+        screen = self.make_screen(input, output)
+        screen.register_palette(PALETTE_COLOR)
+        self.urwid_loop = urwid.MainLoop(
+            self.ui, screen=screen,
+            handle_mouse=False, pop_ups=True,
+            unhandled_input=self.unhandled_input,
+            event_loop=urwid.AsyncioEventLoop(loop=self.aio_loop),
+            **self.extra_urwid_loop_args()
+            )
+        extend_dec_special_charmap()
+        self.toggle_rich()
+        self.urwid_loop.start()
+
+    def initial_controller_index(self):
+        if not self.updated:
+            return 0
+        state_path = self.state_path('last-screen')
+        if not os.path.exists(state_path):
+            return 0
+        with open(state_path) as fp:
+            last_screen = fp.read().strip()
+        controller_index = 0
+        for i, controller in enumerate(self.controllers.instances):
```

Comment:
This loop can be replaced by a call to index, with an except returning zero

Reply:
Not directly it can't? We're searching by name.